### DISCUSSION TITLE
node version http/ws performance

### DISCUSSION THREAD

**Discussion by hst-m** [uNetworking/uWebSockets.js#341] [Q&A] [answered]
ran some benchmarks with [nvm](https://github.com/nvm-sh/nvm#node-version-manager---) and it does appear there is correlation with each new node version and decreasing performance of uWS.js http (requests per second)

![image](https://user-images.githubusercontent.com/65402951/86506880-666f6600-bd88-11ea-9f6b-66b982573273.png)

**Comment by ghost** [comment, marked as answer]
Absolutely. That's what I've been saying for years. The only limit uWS.js has is `node::MakeCallback` and they manage to tamper with that every release, to our disadvantage.

They never benchmark their changes and when they do, they still allow small drops in perf. by every release, because they have no discipline or leadership to refuse offending commits. Many drops make an ocean over time.

**Comment by hst-m** [comment]
Websocket echos per second also showing the same thing. What is going on over there they just keep adding more code and it is slowing stuff down?


![image](https://user-images.githubusercontent.com/65402951/86503815-8f7eff00-bd66-11ea-9432-ce3c04520300.png)

**Comment by hst-m** [comment]
Must be something they can do to fix it, that's a 25% performance drop for sockets and 20% drop for http from node v10 and this library only hooks into theirs minimally so maybe have someone over there look into it, get it back up to 100%

**Comment by ghost** [comment]
If you look at the history of (C++) uWS you see that the trend went the other way:

v0.14.8:
Running benchmark now...
Msg/sec: 199302.500000
Msg/sec: 200806.000000
Msg/sec: 198921.750000
Msg/sec: 198346.000000

v18.9.0:
Running benchmark now...
Msg/sec: 245320.500000
Msg/sec: 246637.500000
Msg/sec: 247706.000000
Msg/sec: 246724.750000

This is for WebSockets but the same can be seen in Http, even though v18 has a complete router on top.

We've remained at a flat line since the jump from v0.14.8 -> v0.15.0, because we do not allow small drops to slowly form an ocean over time.

If someone sends a PR that decrements perf. I refuse it.

**Comment by ghost** [comment]
I wrote a benchmark that just does `node::MakeCallback` a million times in different Node.js versions.

* Node.js 8.0.0 does it in 161ms.
* Node.js 10.0.0 does it in 193ms.
* Node.js 13.8.0 does it in 273ms.
* Node.js 13.9.0 does it in 622ms.
* Node.js 14.5.0 does it in about 600ms, but sometimes 800ms (huge variations).

And this benchmark does nothing at all besides calling `node::MakeCallback` - the single one function we are forced to use.

If you then look at the vanilla (Google) V8 counterpart we get 104ms in Node.js 8, and 136ms in Node.js 14.5.0.

So anything above 200ms is definitely a regression caused by Node.js itself, and not V8 or any native addon.

**Comment by ghost** [comment]
@BridgeAR You had some questions regrading this a while back - what's your take on this?

**Comment by ghost** [comment]
![chart](https://user-images.githubusercontent.com/3397140/86506040-a99ef880-bdcb-11ea-9758-6c8fd340ade0.png)

For the visual learners

**Comment by ghost** [comment]
Alright, nothing we can do here.

**Comment by ghost** [comment]
Apparently @trevnorris stopped working on nodejs core 3 years ago, and I guess @addaleax took over responsibility for those parts. This maps well with the graph; a slow but sure decline in perf. with every release

**Comment by ghost** [comment]
I did manage to reach Trevor Norris on this and he's made a very detail per-commit benchmark and found the most offensive commit in this. That's great but unfortunately Trevor is not working on Node.js anymore, so I don't know if people will listen to him on this.

https://gist.github.com/trevnorris/35c8e570c3097f58739b84bfa3f0a390

**Comment by ghost** [comment]
This is the most offending commit: https://github.com/nodejs/node/pull/30959

It's ironic that they claim to have run benchmarks when they **clearly** did not!

**Comment by hst-m** [comment]
Trevor stats for `MakeCallback` match up with uWS.js benchmark

![image](https://user-images.githubusercontent.com/65402951/87204135-f47fab00-c2b8-11ea-9798-4f0cb59fde56.png)

**Comment by lewispham** [comment]
@hst-m Is this still a problem in node v17? If that's still the case then why shouldn't we create an issue on nodejs repo?

**Reply by ghost** [reply]
> why shouldn't we create an issue on nodejs repo

Because they don't care.

**Reply by lewispham** [reply]
But as we're having evidence then we should still have an opening issue for them right? It'll surely be difficult as uws.js is trying to compete with nodejs core http lib but I think it shouldn't be the reason to avoid working with them to resolve the issue.

**Reply by e3dio** [reply]
I think performance improved right after this in Node 15, they worked on updating this. There was direct correlation with MakeCallback and uWS performance as you can see in graph but they made updates. Would need to test again to compare up to v17

**Reply by digitalml** [reply]
@hst-m any chance of getting some new benchmarks with v15 through v17 please?

**Reply by ronag** [reply]
@lewispham Feel free to create an issue on the node repo. Explain the concern and give some benchmark numbers. I suspect some of the overhead is due to async hooks. Maybe ask if there is a possibility to disable this for those that don't care about async context and just want maximum performance.

I believe Alex has created issues for this in the past so I don't know if it's likely it to be changed. But it was quite some time ago and I don't think it would hurt to have another go at discussing it.

I tried taking a look at it today but it's unfortunately out of my depth.

**Reply by ronag** [reply]
Yes, I have noticed some communication challenges. I'll have go at your patch and see what I can do. Thanks.

**Reply by ronag** [reply]
https://github.com/nodejs/node/pull/41279

**Reply by abenhamdine** [reply]
For information, last nodejs version https://github.com/nodejs/node/releases/tag/v17.4.0 includes the perf improvement implemented in https://github.com/nodejs/node/pull/41331 which I think adress that issue.

**Comment by e3dio** [comment]
I ran benchmarks back to Node 14, I see 18% improvement since this issue was created, and the [commit](https://github.com/nodejs/node/pull/41331) at 17.4 helped. The lowest point was Node 14.5, and the latest Node 17.4 is the fastest since Node 12. It's about half way back to the Node 10 high. So it's good to see a positive direction

![Slide-1-copy-1 (1)](https://user-images.githubusercontent.com/85405955/151994418-0f042365-beab-44cf-a9ed-6c94f41ea9d7.jpg)

**Reply by ghost** [reply]
Nice graph! Hard to be excited about a **less degraded** performance though.

**Reply by e3dio** [reply]
Yes just following up for these people
> any chance of getting some new benchmarks with v15 through v17 please? @digitalml

> this still a problem in node v17? @lewispham

I was half expecting to see performance continue a nosedive in the downward direction, but was pleasantly surprised to see it reverse course

**Reply by abenhamdine** [reply]
> I was half expecting to see performance continue a nosedive in the downward direction, but was pleasantly surprised to see it reverse course

Personnaly I'm a bit disappointed by these figures, I was expecting from https://github.com/nodejs/node/pull/41331 a greater improvement than 3%

**Reply by ghost** [reply]
You can't polish a turd

**Reply by abenhamdine** [reply]
The main problem in Nodejs concerning performance issues like this one is that there no automatic performance regression checks, the benchmark WG has been decommissionned, and there's only one core contributor, mscdex, who is actively tracking performances issues/regressions.

**Reply by ghost** [reply]
It is what it is
